File: fork-runner/src/main/java/com/shazam/fork/summary/executiontimeline/StatSummarySerializer.java
package com.shazam.fork.summary.executiontimeline;

import com.agoda.fork.stat.TestHistory;
import com.agoda.fork.stat.TestMetric;
import com.android.ddmlib.testrunner.TestIdentifier;
import com.shazam.fork.model.Device;
import com.shazam.fork.stat.TestExecution;
import com.shazam.fork.stat.TestExecutionReporter;
import com.shazam.fork.stat.TestHistoryManager;
import com.shazam.fork.stat.TestStatsLoader;
import com.shazam.fork.summary.PoolSummary;
import com.shazam.fork.summary.Summary;
import com.shazam.fork.summary.TestResult;

import java.util.List;
import java.util.stream.Stream;

import static java.util.stream.Collectors.toList;

public class StatSummarySerializer {

    private final TestExecutionReporter reporter;
    private final TestStatsLoader testStatsLoader;

    public StatSummarySerializer(TestExecutionReporter reporter,
                                 TestStatsLoader testStatsLoader) {
        this.reporter = reporter;
        this.testStatsLoader = testStatsLoader;
    }

    private String prepareTestName(String fullTestName) {
        return fullTestName.substring(fullTestName.lastIndexOf('.') + 1);
    }

    private Stream<Data> parseData(Device device) {
        List<TestExecution> executions = reporter.getTests(device);
        return executions.stream().map(this::convertToData);
    }

    private Data convertToData(TestExecution execution) {
        int status = convertStatus(execution);
        String preparedTestName = prepareTestName(execution.getTest().toString());
        TestMetric testMetric = getTestMetric(execution);
        return createData(execution, status, preparedTestName, testMetric);
    }

    private int convertStatus(TestExecution execution) {
        return execution.getStatus() == TestExecution.Status.ENDED ? 1 : 0;
    }

    private Data createData(TestExecution execution, int status, String preparedTestName, TestMetric testMetric) {
        return new Data(preparedTestName, status,
                execution.getStartTime(),
                execution.getStartTime() + execution.getEndTime(),
                testMetric.getExpectedValue(), testMetric.getVariance());
    }

    private TestMetric getTestMetric(TestExecution execution) {
        return testStatsLoader.findMetric(execution.getTest());
    }

    private ExecutionStats calculateExecutionStats(List<Data> data) {

        return new ExecutionStats(calculateIdle(data), calculateAverageExecutionTime(data));
    }

    private long calculateAverageExecutionTime(List<Data> data) {
        return (long) data.stream()
                .mapToLong(this::calculateDuration)
                .average()
                .orElse(0.0);
    }

    private long calculateDuration(Data a) {
        return a.getEndDate() - a.getStartDate();
    }

    private Long calculateIdle(List<Data> data) {
        if (data.size() == 0) return 0L;
        long time = 0;
        for (int i = 1; i < data.size(); i++) {
            time += data.get(i).getStartDate() - data.get(i - 1).getEndDate();
        }
        return time;
    }

    private Measure createMeasure(Device device) {
        List<Data> data = parseData(device).collect(toList());
        return new Measure(device.getSerial(), calculateExecutionStats(data), data);
    }

    private Stream<Measure> parsePoolSummary(PoolSummary poolSummary) {
        return poolSummary.getTestResults()
                .stream()
                .map(TestResult::getDevice)
                .distinct()
                .map(this::createMeasure);
    }


    private List<Measure> parseList(List<PoolSummary> poolSummaries) {
        return poolSummaries.stream().flatMap(this::parsePoolSummary).collect(toList());
    }

    private Stream<TestIdentifier> extractIdentifiers(PoolSummary summary) {
        return summary.getTestResults()
                .stream()
                .map(result -> new TestIdentifier(result.getTestClass(), result.getTestMethod()));
    }

    private int passedTestCount(Summary summary) {
        return (int) summary.getPoolSummaries()
                .stream()
                .flatMap(this::extractIdentifiers)
                .distinct()
                .count();
    }

    private ExecutionStats aggregateExecutionStats(List<Measure> list) {
        long summaryIdle = 0;
        long avgTestExecutionTime = 0;
        for (Measure measure : list) {
            summaryIdle += measure.getExecutionStats().getIdleTimeMillis();
            avgTestExecutionTime += measure.getExecutionStats().getAverageTestExecutionTimeMillis();
        }
        int size = list.size();
        //Temp workaround for Parameterized tests
        if (size != 0) {
            avgTestExecutionTime = avgTestExecutionTime / size;
        }
        return new ExecutionStats(summaryIdle, avgTestExecutionTime);
    }

    public ExecutionResult parse(Summary summary) {
        int failedTests = summary.getFailedTests().size();
        int ignoredTests = summary.getIgnoredTests().size();
        int passedTestCount = passedTestCount(summary) - ignoredTests - failedTests;
        List<Measure> measures = parseList(summary.getPoolSummaries());
        return new ExecutionResult(passedTestCount, failedTests, aggregateExecutionStats(measures), measures);
    }
}

File: fork-runner/src/main/java/com/shazam/fork/summary/executiontimeline/StatSummarySerializer.kt
package com.shazam.fork.summary.executiontimeline

import com.agoda.fork.stat.TestMetric
import com.android.ddmlib.testrunner.TestIdentifier
import com.shazam.fork.model.Device
import com.shazam.fork.stat.TestExecution
import com.shazam.fork.stat.TestExecutionReporter
import com.shazam.fork.stat.TestStatsLoader
import com.shazam.fork.summary.PoolSummary
import com.shazam.fork.summary.Summary

class StatSummarySerializer(private val reporter: TestExecutionReporter,
                            private val testStatsLoader: TestStatsLoader) {

    private fun prepareTestName(fullTestName: String): String {
        return fullTestName.substring(fullTestName.lastIndexOf('.') + 1)
    }

    private fun parseData(pool: String, device: Device): List<Data> {
        val executions = reporter.getTests(pool,device)
        return executions.map { this.convertToData(it) }
    }

    private fun convertToData(execution: TestExecution): Data {
        val preparedTestName = prepareTestName(execution.test.toString())
        val testMetric = getTestMetric(execution)
        return createData(execution, execution.status, preparedTestName, testMetric)
    }

    private fun createData(execution: TestExecution, status: TestExecution.Status, preparedTestName: String, testMetric: TestMetric): Data {
        return Data(preparedTestName, status,
                execution.startTime,
                execution.startTime + execution.endTime,
                testMetric.expectedValue, testMetric.variance)
    }

    private fun getTestMetric(execution: TestExecution): TestMetric {
        return testStatsLoader.findMetric(execution.test)
    }

    private fun calculateExecutionStats(data: List<Data>): ExecutionStats {
        return ExecutionStats(calculateIdle(data), calculateAverageExecutionTime(data))
    }

    private fun calculateAverageExecutionTime(data: List<Data>): Long {
        return data.map { this.calculateDuration(it) }.average().toLong()
    }

    private fun calculateDuration(a: Data): Long {
        return a.endDate - a.startDate
    }

    private fun calculateIdle(data: List<Data>): Long {
        return data.stream().collect(SlidingCollector(2, 1)).map {
            if (it.size == 2) {
                it[1].startDate - it[0].endDate
            } else {
                0
            }
        }.sum()
    }

    private fun parsePoolSummary(poolSummary: PoolSummary): List<Measure> {
        return poolSummary.testResults
                .map { it.device }
                .distinct()
                .map { this.createMeasure(poolSummary.poolName, it) }
    }

    private fun createMeasure(pool: String, device: Device): Measure {
        val data = parseData(pool, device)
        return Measure(device.serial, calculateExecutionStats(data), data)
    }

    private fun parseList(poolSummaries: List<PoolSummary>): List<Measure> {
        return poolSummaries.flatMap { this.parsePoolSummary(it) }
    }

    private fun extractIdentifiers(summary: PoolSummary): List<TestIdentifier> {
        return summary.testResults
                .map { result -> TestIdentifier(result.testClass, result.testMethod) }
    }

    private fun passedTestCount(summary: Summary): Int {
        return summary.poolSummaries
                .flatMap { this.extractIdentifiers(it) }
                .distinct()
                .count()
    }

    private fun aggregateExecutionStats(list: List<Measure>): ExecutionStats {
        val summaryIdle = list
                .map { it.executionStats.idleTimeMillis }
                .sum()
        val avgTestExecutionTime = list
                .map { it.executionStats.averageTestExecutionTimeMillis }
                .average()
                .toLong()
        return ExecutionStats(summaryIdle, avgTestExecutionTime)
    }

    fun parse(summary: Summary): ExecutionResult {
        val failedTests = summary.failedTests.size
        val ignoredTests = summary.ignoredTests.size
        val passedTestCount = passedTestCount(summary)
        val measures = parseList(summary.poolSummaries)
        return ExecutionResult(passedTestCount, failedTests, aggregateExecutionStats(measures), measures)
    }
}
